File: julia/Other/advent_of_code/2022/day1/day1.jl
parse_input(input_file::String) =
    [[parse(Int, i) for i in split(s)] for s in split(read(input_file, String), "\n\n")]

part1(data::Vector{Vector{Int}}) =
    maximum(sum(d) for d in data)

part2(data::Vector{Vector{Int}}) =
    sum(sort([sum(d) for d in data], rev=true)[1:3])

function main()
    data = parse_input("data1.txt")

    # Part 1
    part1_solution = part1(data)
    println("Part 1: $part1_solution")

    # Part 2
    part2_solution = part2(data)
    println("Part 2: $part2_solution")
end

main()

File: julia/Other/advent_of_code/2021/day1/day1.hs
main :: IO ()
main = do
    eachLine <- lines <$> readFile "data1.txt"
    let nums = map read eachLine :: [Int]
    
    let part1Solution = part1 nums
    putStrLn $ "Part 1: " ++ show part1Solution

    let part2Solution = part2 nums
    putStrLn $ "Part 2: " ++ show part2Solution


withLaggedPairs :: Int -> (a -> a -> b) -> [a] -> [b]
withLaggedPairs n f xs = zipWith f xs (drop n xs)

laggedPairs :: Int -> [a] -> [(a,a)]
laggedPairs n = withLaggedPairs n (,)

countTrue p = length . filter p

countIncreases n = countTrue (uncurry (<)) . laggedPairs n


part1 :: [Int] -> Int
-- part1 nums = head [ | i <- nums, j <- nums, i + j == 2020]
part1 = countIncreases 1

part2 :: [Int] -> Int
part2 = countIncreases 3
